### PR TITLE
Add flash message with info after login

### PR DESCRIPTION
**Pull request by Mordorreal**
- Add new types for flash message
- Show info type flash message after login via github
- Add missed tests

Small initial step to address #19

### CODE REVIEW

File: app/routes/login.js [hunk 1]
```diff
@@ -67,6 +68,12 @@ export default Route.extend({
       let user = this.store.push(this.store.normalize('user', data.user));
       let transition = this.get('session.savedTransition');
       this.session.loginUser(user);
+
+      const messageAfterLogin = htmlSafe(
+        "Welcome to crates.io! Visit <a href='me'>account settings</a> to verify your email address and create an API token!",
+      );
+      this.flashMessages.show(messageAfterLogin, { type: 'info' });
```
Comment (Turbo87) [inline app/routes/login.js]: do you want to show this message to **all** users? even if they have verified their data already? 🤔

Comment (Mordorreal) [reply app/routes/login.js]: For first version yes. Currently trying  to understand how I can pass data from backend and what parameters it will be. I think about `{ "email_verified": false, "api_key_verified": true }` but backed currently is a real mess for me. No good structure and abstractions after rails. Still trying to learn how to handle that.

Comment (carols10cents) [reply app/routes/login.js]: Hi @mordorreal! Thank you for this pull request!

I'm sorry our backend is such a mess; there isn't really a Rails for Rust yet and crates.io was written even before any of the experimental web frameworks existed!

There is a [router.rs file](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs) that is similar (in purpose at least) to Rails' routes.rb. There are two backend routes that will be relevant for this PR:

- [`/api/private/session/authorize`](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs#L118) is how the frontend gets information about the current user from the backend immediately after logging in with GitHub. It [calls](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs#L119) the [`user::session::authorize` controller function](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/controllers/user/session.rs#L46-L105). After interacting with info from GitHub, it [calls](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/controllers/user/session.rs#L104) the `user::me::me` controller function (discussed more next)
- [`/api/v1/me`](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs#L79) is how the frontend gets information about the current user from the backend if there's already an existing session for a logged in user. It also calls the [`user::me::me` controller function](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/controllers/user/me.rs#L14-L61), so this URL returns the same data as the session authorize URL.

The data returned currently is defined by the "view" struct [`EncodableMe`](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/views.rs#L159-L163). Because the backend is only a JSON API, there are no HTML views, so all the views are structs that get serialized to JSON with serde.

`EncodableMe` contains an [`EncodablePrivateUser`](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/views.rs#L165-L178), which has an [`email_verified` field](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/views.rs#L172), so you should be able to use that!

We can add a field to `EncodablePrivateUser` called `has_tokens` or similar that is `true` if the user has at least one token and `false` if they have none. We'd need to add a query to `user::me::me` that gets that information. The query should use `ApiToken::belonging_to` sort of like [this code does](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/controllers/token.rs#L13-L14) and diesel's `exists` function sort of like [this code does](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/models/krate.rs#L163-L169). I'm actually not sure off the top of my head exactly how to combine those two! [Diesel's docs](http://docs.diesel.rs/) is where I'd look to figure that out. Want to give that a try? 

Let me know if you have more questions, I hope that helps to pick the relevant pieces out of our mess at least!

Comment (Mordorreal) [reply app/routes/login.js]: Hi @carols10cents! Thank you for very detailed answer! You're awesome! I'll try to continue on the issue in this PR.

**Comment by Mordorreal** [comment]
@carols10cents Update with all fixes and improvements. Feedback is welcome!

**Comment by Mordorreal** [comment]
Can someone with permissions also restart CI? Looks like there some problems with testing environment.

File: app/routes/me/index.js [hunk 2]
```diff
@@ -5,12 +5,10 @@ import AuthenticatedRoute from '../../mixins/authenticated-route';
...
-        .setProperties({
-          emailNotificationsSuccess: false,
-          emailNotificationsError: false,
-        })
-        .clear();
+      this.controller.setProperties({
+        emailNotificationsSuccess: false,
+        emailNotificationsError: false,
+      });
     },
```
Comment (Turbo87) [inline app/routes/me/index.js]: why did you change the formatting here?

File: app/templates/catch-all.hbs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<p id='flash' class='shown'>Oops, that route doesn't exist!</p>
+<p id='flash' class='show'>Oops, that route doesn't exist!</p>
```
Comment (Turbo87) [inline app/templates/catch-all.hbs]: why did you change the class name?

File: tests/integration/components/welcome-message-test.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  test('it show reminder about email only if user has tokens', async function(assert) {
+    assert.expect(2);
+    const user = this.server.create('user', 'withTokens');
+
+    this.session = this.owner.lookup('service:session');
+    this.session.loginUser(user);
+
+    await render(hbs`{{welcome-message}}`);
+
+    assert.equal(
```
Comment (Turbo87) [inline tests/integration/components/welcome-message-test.js]: these checks should use qunit-dom assertions

File: app/services/flash-messages.js [hunk 5]
```diff
@@ -4,8 +4,9 @@ export default Service.extend({
   message: null,
   _nextMessage: null,
 
-  show(message) {
+  show(message, options = { type: 'warning' }) {
     this.set('message', message);
+    this.set('options', options);
   },
```
Comment (Turbo87) [inline app/services/flash-messages.js]: are these changes still necessary? it looks like we're not using flash messages anymore

File: app/routes/login.js [hunk 6]
```diff
@@ -68,6 +68,7 @@ export default Route.extend({
       let user = this.store.push(this.store.normalize('user', data.user));
       let transition = this.get('session.savedTransition');
       this.session.loginUser(user);
+
```
Comment (Turbo87) [inline app/routes/login.js]: ?

File: app/components/welcome-message.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    const textArray = [
+      !user.email_verified && 'verify your email address',
+      !user.email_verified && !user.has_tokens && ' and ',
+      !user.has_tokens && 'create an API token',
+      '!',
+    ].filter(e => !!e);
+
+    return textArray.join('');
+  }),
```
Comment (Turbo87) [inline app/components/welcome-message.js]: I think I would prefer it if we moved that code into the index controller and not instantiate the component at all if we don't have anything to display

File: app/components/welcome-message.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  text: computed('session.currentUser.{email_verified,has_tokens}', function() {
+    const user = this.get('session.currentUser');
+    if (!user || (user.email_verified && user.has_tokens)) return '';
+
+    const textArray = [
+      !user.email_verified && 'verify your email address',
+      !user.email_verified && !user.has_tokens && ' and ',
+      !user.has_tokens && 'create an API token',
+      '!',
+    ].filter(e => !!e);
```
Comment (Turbo87) [inline app/components/welcome-message.js]: the code logic here seems a bit hard to read. do you think it could be simplified in some way?

File: app/routes/me/index.js (hunk 2 above)
Comment (Mordorreal) [reply app/routes/me/index.js]: There was an error in console. Please, see commit comment.

File: app/templates/catch-all.hbs (hunk 3 above)
Comment (Mordorreal) [reply app/templates/catch-all.hbs]: For me, the noun is made more sense here, because this is what this css class doing. But English is not my native language and maybe I'm wrong.

File: tests/integration/components/welcome-message-test.js (hunk 4 above)
Comment (Mordorreal) [reply tests/integration/components/welcome-message-test.js]: ok. thanks. Will fix.

File: app/services/flash-messages.js (hunk 5 above)
Comment (Mordorreal) [reply app/services/flash-messages.js]: If the project has no plan to bring better user exp with right info or success messages then I can remove it.

File: app/routes/login.js (hunk 6 above)
Comment (Mordorreal) [reply app/routes/login.js]: Nice catch. will remove.

File: app/components/welcome-message.js (hunk 7 above)
Comment (Mordorreal) [reply app/components/welcome-message.js]: This logic has nothing in common with the index controller. This is common arch mistake in all MVC frameworks. Trying to move business logic inside of controllers fail by design.

File: app/components/welcome-message.js (hunk 8 above)
Comment (Mordorreal) [reply app/components/welcome-message.js]: I tried some variants before come to this. For example, a variant with ternary operators looks more ugly:
```
const verifyText = user.email_verified ? '' : 'verify your email address ';
const andText = user.email_verified && user.has_tokens ? '' : 'and ';
const tokenText = user.has_tokens ? '' : 'create an API token';

return `${verifyText}${andText}{$tokenText}`;
```
Moving this calculation logic to template is not right.
I'm open to any ideas on how to improve this code.

File: tests/integration/components/welcome-message-test.js (hunk 4 above)
Comment (Mordorreal) [reply tests/integration/components/welcome-message-test.js]: Done.

File: app/routes/login.js (hunk 6 above)
Comment (Mordorreal) [reply app/routes/login.js]: Done

**Comment by Mordorreal** [comment]
@carols10cents Sorry for disturbing but can you please review the pr?

**Comment by Mordorreal** [comment]
@carols10cents, @Turbo87 Hi. Can you please review? I don't know why label is saying that pr waiting for the author because it's not true. PR is waiting for review.

**Comment by locks** [comment]
@Mordorreal you have multiple conflicts at the moment. Can you rebase and resolve them?

**Comment by Mordorreal** [comment]
@locks Ohhhh. Rebased. That was hard. Can you please review pr? Not sure that I will have willing to rebase this pr again after one month of silence.

**Comment by Mordorreal** [comment]
@Turbo87 Are this PR still needed? If no, please close. If yes, I have time for it now.

**Comment by Turbo87** [comment]
closing due to https://github.com/rust-lang/crates.io/issues/19#issuecomment-1914335696 :)
